bottomless/src/replicator.rs

```diff
@@ -620,19 +620,18 @@ impl Replicator {
                 Self::read_change_counter(&mut reader).await?
             }
             CompressionKind::Gzip => {
-                // TODO: find a way to compress ByteStream on the fly instead of creating
-                // an intermediary file.
-                let (compressed_db_path, change_counter) = self.compress_main_db_file().await?;
+                let mut reader = tokio::fs::File::open(&self.db_path).await?;
+                let buf_reader = tokio::io::BufReader::new(reader.try_clone().await?);
+                let gzip_reader = async_compression::tokio::bufread::GzipEncoder::new(buf_reader);
+
                 let key = format!("{}-{}/db.gz", self.db_name, self.generation);
-                self.client
-                    .put_object()
-                    .bucket(&self.bucket)
-                    .key(key)
-                    .body(ByteStream::from_path(compressed_db_path).await?)
-                    .send()
-                    .await?;
-                let _ = tokio::fs::remove_file(compressed_db_path).await;
-                change_counter
+
+                // Unfortunally we can't send the gzip output in a single call without buffering
+                // the whole snapshot in memory because S3 requires the `Content-Length` header
+                // to be set.
+                upload_s3_multipart(&self.client, &key, &self.bucket, gzip_reader).await?;
+
+                Self::read_change_counter(&mut reader).await?
```

Comment:
nitpick: this line can be deduplicated from line 620 and just moved out of the `match` clause, along with the `mut reader = tokio::fs::File::open` part

Reply:
done :)